PYIC-7632: Include kid in private key jwt for token exchange

Wynndow opened the pull request:
<!-- Provide a general summary of your changes in the Title above -->
<!-- Include the Jira ticket number in square brackets as prefix, eg `[P4-XXXX] PR Title` -->

## Proposed changes

### What changed

Include kid in private key jwt for token exchange

### Why did it change

When we exchange an auth code for a token with a CRI we use a signed JWT to prove the request is coming from a legitimate source. This change updates the header of the JWT to include the kid of the signing key.

We already had logic to include the key when signing - it was being used when generating a JAR. It was implemented for DWP and is behind a flag to allow it to be tested in staging before hitting prod.

This moves the logic to a shared place and utilises it for both signing operations.

### Issue tracking
<!-- List any related Jira tickets or GitHub issues -->
<!-- List any related ADRs or RFCs -->
<!-- Delete/copy as appropriate -->

- [PYIC-7632](https://govukverify.atlassian.net/browse/PYIC-7632)


[PYIC-7632]: https://govukverify.atlassian.net/browse/PYIC-7632?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

Wynndow committed:
PYIC-7632: Include kid in private key jwt for token exchange

When we exchange an auth code for a token with a CRI we use a signed JWT
to prove the request is coming from a legitimate source. This change
updates the header of the JWT to include the kid of the signing key.

We already had logic to include the key when signing - it was being used
when generating a JAR. It was implemented for DWP and is behind a flag
to allow it to be tested in staging before hitting prod.

This moves the logic to a shared place and utilises it for both signing
operations.

Joe-Edwards-GDS commented on libs/common-services/src/main/java/uk/gov/di/ipv/core/library/helpers/JwtHelper.java:
```diff
@@ -8,38 +8,42 @@
-    public static <T> SignedJWT createSignedJwtFromObject(T claimInput, JWSSigner signer)
-            throws JOSEException {
-        JWSHeader jwsHeader = generateHeader();
+    public static <T> SignedJWT createSignedJwtFromObject(
```
I wonder if we should change this signature to take a `JWTClaimsSet` instead, and get rid of `generateClaims`...

Looking at the other usage (for the private key JWT), we'd be better served getting rid of our `ClientAuthClaims` object and using the built-in nimbus classes. E.g.:
```
            var clientAuthClaims = new JWTClaimsSet.Builder()
                    .issuer(criConfig.getClientId())
                    .subject(criConfig.getClientId())
                    .audience(criConfig.getComponentId())
                    .expirationTime(...)
                    .jwtID(secureTokenHelper.generate())
                    .build();
```
IMO this is clearer and safer than using our custom class and serialising ourselves! And it means you can use a shared method for the signing.

Appreciate it's nothing really to do with your change, so happy to put into a different PR if you prefer.

Wynndow replied:
Yep, that makes sense. I'll fix it up here as I'm in that headspace.

Joe-Edwards-GDS commented on libs/common-services/src/main/java/uk/gov/di/ipv/core/library/helpers/JwtHelper.java:
```diff
@@ -8,38 +8,42 @@
-            throws JOSEException {
-        JWSHeader jwsHeader = generateHeader();
+    public static <T> SignedJWT createSignedJwtFromObject(
+            T claimInput, JWSSigner signer, ConfigService configService) throws JOSEException {
```
I don't _love_ that this takes a `configService` - it feels like a bad abstraction for a static helper class, but can live with it if there's no neater option. (I might prefer to have `includeKeyId` as a parameter, perhaps)

Wynndow replied:
Yeah, we could pass the keyId in, and then include it if it's there.

Joe-Edwards-GDS commented on libs/common-services/src/main/java/uk/gov/di/ipv/core/library/helpers/JwtHelper.java:
```diff
@@ -58,4 +62,26 @@ public static boolean signatureIsDerFormat(SignedJWT signedJWT) throws JOSEExcep
                 != ECDSA.getSignatureByteArrayLength(JWSAlgorithm.ES256);
     }
+
+    private static String getKid(ConfigService configService) {
```
It's annoying that you can't get this from the `JWSSigner` - it feels like an omission :(

Means that this must be kept in sync with the logic in `SignerFactory`, to make sure the key id actually matches the key being used in the signer!

Probably not the end of the world, but a potential improvement might be to either wrap in our own class (that does expose the key id) or combine this helper class with the signer factory, to have a single 'signing service' that handles the key selection in one place.

Joe-Edwards-GDS commented on libs/common-services/src/main/java/uk/gov/di/ipv/core/library/helpers/JwtHelper.java:
```diff
@@ -8,38 +8,42 @@
-public class JwtHelper {
+import static uk.gov.di.ipv.core.library.config.ConfigurationVariable.SIGNING_KEY_ID;
+import static uk.gov.di.ipv.core.library.config.ConfigurationVariable.SIGNING_KEY_JWK;
+import static uk.gov.di.ipv.core.library.config.CoreFeatureFlag.KID_JAR_HEADER;
```
I hadn't realised this was already behind a feature flag! Convenient - means we can test both headers at the same time :D

Joe-Edwards-GDS reviewed (commented):
Looks good but left a few comments about possible improvements :)

Wynndow committed:
PYIC-7632: Refactor to improve reuseability

This refactors the signing logic so that we can share the same helpers
to do all the signing in both places that we sign JWTs.

It also ensures that the key used to create a signer is available with
the signer - this way we can be sure that the kid being used in the
header of a signed JWT is linked to the signer being used to sign.

This moves the kms signer lib we had into the common-services lib, as
they now depend on each other.

Joe-Edwards-GDS commented on libs/common-services/src/main/java/uk/gov/di/ipv/core/library/helpers/JwtHelper.java:
```diff
@@ -1,48 +1,30 @@
     }
 
-    public static JWSHeader generateHeader(ConfigService configService) {
+    public static JWSHeader generateHeader(String kid) {
```
(could now be private)

Wynndow replied:
yep, will update

Joe-Edwards-GDS commented on libs/common-services/src/test/java/uk/gov/di/ipv/core/library/fixtures/TestFixtures.java:
```diff
@@ -18,12 +18,12 @@ public interface TestFixtures {
             "MIGHAgEAMBMGByqGSM49AgEGCCqGSM49AwEHBG0wawIBAQQgOXt0P05ZsQcK7eYusgIPsqZdaBCIJiW4imwUtnaAthWhRANCAAQT1nO46ipxVTilUH2umZPN7OPI49GU6Y8YkcqLxFKUgypUzGbYR2VJGM+QJXk0PI339EyYkt6tjgfS+RcOMQNO"; // pragma: allowlist secret
     String EC_PRIVATE_KEY_JWK =
-            "{\"kty\":\"EC\",\"d\":\"OXt0P05ZsQcK7eYusgIPsqZdaBCIJiW4imwUtnaAthU\",\"crv\":\"P-256\",\"x\":\"E9ZzuOoqcVU4pVB9rpmTzezjyOPRlOmPGJHKi8RSlIM\",\"y\":\"KlTMZthHZUkYz5AleTQ8jff0TJiS3q2OB9L5Fw4xA04\"}"; // pragma: allowlist secret
+            "{\"kty\":\"EC\",\"kid\":\"test-fixtures-ec-key\",\"use\":\"sig\",\"d\":\"OXt0P05ZsQcK7eYusgIPsqZdaBCIJiW4imwUtnaAthU\",\"crv\":\"P-256\",\"x\":\"E9ZzuOoqcVU4pVB9rpmTzezjyOPRlOmPGJHKi8RSlIM\",\"y\":\"KlTMZthHZUkYz5AleTQ8jff0TJiS3q2OB9L5Fw4xA04\"}"; // pragma: allowlist secret
```
(FYI @thebauSoftwire this will probably clash)

Joe-Edwards-GDS reviewed (dismissed):
Thanks - I think that's improved things!
